File: backend/src/plugins/token_handler.py
"""
Token Handler Plugin for PinkPay Payment Switch
Manages offline payment tokens for situations with poor connectivity
"""

import asyncio
import secrets
from datetime import datetime, timedelta
from typing import Dict, Any, Optional
from .plugin_manager import BasePlugin

class TokenHandlerPlugin(BasePlugin):
    """Token Handler Plugin for offline payment tokens"""
    
    def __init__(self):
        super().__init__()
        self.name = "Token Handler"
        self.version = "1.0.0"
        self.description = "Offline payment token management and redemption"
        self.critical = False
        
        self.config = {
            'token_expiry_hours': 24,
            'max_token_amount': 1000.0,  # MYR
            'min_token_amount': 5.0,     # MYR
            'max_tokens_per_user': 5,
            'supported_currencies': ['MYR', 'USD', 'SGD'],
            'auto_cleanup_expired': True
        }
    
    async def execute(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Execute token handling logic"""
        try:
            operation = data.get('token_operation', 'validate')
            
            if operation == 'create':
                return await self._create_token(data)
            elif operation == 'redeem':
                return await self._redeem_token(data)
            elif operation == 'validate':
                return await self._validate_token(data)
            elif operation == 'cancel':
                return await self._cancel_token(data)
            else:
                return {
                    'success': True,
                    'data': {
                        'token_operation': 'none',
                        'token_required': False
                    }
                }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Token handling failed: {str(e)}",
                'critical': False
            }
    
    async def _create_token(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new offline token"""
        try:
            user_id = data.get('user_id')
            amount = float(data.get('amount', 0))
            currency = data.get('currency', 'MYR')
            
            # Validate token creation
            validation_result = self._validate_token_creation(amount, currency, user_id)
            if not validation_result['valid']:
                return {
                    'success': False,
                    'error': validation_result['reason'],
                    'critical': False
                }
            
            # Generate token
            token_data = await self._generate_offline_token(user_id, amount, currency)
            
            # In production, this would save to database
            # For demo, we'll simulate the database operation
            await asyncio.sleep(0.1)
            
            return {
                'success': True,
                'data': {
                    'token_created': True,
                    'token_data': token_data,
                    'token_operation': 'create'
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Token creation failed: {str(e)}",
                'critical': False
            }
    
    async def _redeem_token(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Redeem an offline token"""
        try:
            token_id = data.get('token_id')
            
            if not token_id:
                return {
                    'success': False,
                    'error': "Token ID required for redemption",
                    'critical': False
                }
            
            # Validate and redeem token
            redemption_result = await self._validate_and_redeem_token(token_id)
            
            return {
                'success': True,
                'data': {
                    'token_redeemed': redemption_result['valid'],
                    'redemption_data': redemption_result,
                    'token_operation': 'redeem'
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Token redemption failed: {str(e)}",
                'critical': False
            }
    
    async def _validate_token(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Validate an offline token without redeeming"""
        try:
            token_id = data.get('token_id')
            
            if not token_id:
                return {
                    'success': True,
                    'data': {
                        'token_valid': False,
                        'token_operation': 'validate',
                        'reason': 'No token provided'
                    }
                }
            
            # Validate token
            validation_result = await self._check_token_validity(token_id)
            
            return {
                'success': True,
                'data': {
                    'token_valid': validation_result['valid'],
                    'validation_data': validation_result,
                    'token_operation': 'validate'
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Token validation failed: {str(e)}",
                'critical': False
            }
    
    async def _cancel_token(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Cancel an offline token"""
        try:
            token_id = data.get('token_id')
            user_id = data.get('user_id')
            
            # Validate cancellation
            cancellation_result = await self._cancel_user_token(token_id, user_id)
            
            return {
                'success': True,
                'data': {
                    'token_cancelled': cancellation_result['cancelled'],
                    'cancellation_data': cancellation_result,
                    'token_operation': 'cancel'
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Token cancellation failed: {str(e)}",
                'critical': False
            }
    
    async def _generate_offline_token(self, user_id: str, amount: float, currency: str) -> Dict[str, Any]:
        """Generate a new offline token"""
        
        # Generate secure token ID
        token_id = self._generate_token_id()
        
        # Calculate expiry
        expires_at = datetime.utcnow() + timedelta(hours=self.config['token_expiry_hours'])
        
        # Create token data
        token_data = {
            'token_id': token_id,
            'user_id': user_id,
            'amount': amount,
            'currency': currency,
            'status': 'active',
            'created_at': datetime.utcnow().isoformat(),
            'expires_at': expires_at.isoformat(),
            'creation_location': 'unknown',  # Could be enhanced with geolocation
            'qr_code': self._generate_token_qr_code(token_id)
        }
        
        return token_data
    
    async def _validate_and_redeem_token(self, token_id: str) -> Dict[str, Any]:
        """Validate and redeem a token"""
        
        # Simulate database lookup
        await asyncio.sleep(0.1)
        
        # Mock token data (in production, fetch from database)
        mock_tokens = {
            'TOKEN_12345': {
                'token_id': 'TOKEN_12345',
                'user_id': 'user_001',
                'amount': 50.0,
                'currency': 'MYR',
                'status': 'active',
                'expires_at': (datetime.utcnow() + timedelta(hours=12)).isoformat()
            }
        }
        
        token_data = mock_tokens.get(token_id)
        
        if not token_data:
            return {
                'valid': False,
                'reason': 'Token not found',
                'redeemed': False
            }
        
        # Check if token is still valid
        if token_data['status'] != 'active':
            return {
                'valid': False,
                'reason': f"Token status is {token_data['status']}",
                'redeemed': False
            }
        
        # Check expiry
        expires_at = datetime.fromisoformat(token_data['expires_at'].replace('Z', '+00:00'))
        if datetime.utcnow() > expires_at.replace(tzinfo=None):
            return {
                'valid': False,
                'reason': 'Token has expired',
                'redeemed': False
            }
        
        # Token is valid, mark as redeemed
        token_data['status'] = 'redeemed'
        token_data['redeemed_at'] = datetime.utcnow().isoformat()
        
        return {
            'valid': True,
            'redeemed': True,
            'token_data': token_data,
            'redemption_timestamp': datetime.utcnow().isoformat()
        }
    
    async def _check_token_validity(self, token_id: str) -> Dict[str, Any]:
        """Check if a token is valid without redeeming it"""
        
        # Simulate database lookup
        await asyncio.sleep(0.05)
        
        # Mock validation (in production, query database)
        if token_id.startswith('TOKEN_'):
            return {
                'valid': True,
                'reason': 'Token is valid and active',
                'amount': 50.0,
                'currency': 'MYR',
                'expires_in_hours': 12
            }
        else:
            return {
                'valid': False,
                'reason': 'Invalid token format'
            }
    
    async def _cancel_user_token(self, token_id: str, user_id: str) -> Dict[str, Any]:
        """Cancel a user's token"""
        
        # Simulate database operation
        await asyncio.sleep(0.1)
        
        # Mock cancellation logic
        if token_id and user_id:
            return {
                'cancelled': True,
                'reason': 'Token cancelled successfully',
                'cancellation_timestamp': datetime.utcnow().isoformat()
            }
        else:
            return {
                'cancelled': False,
                'reason': 'Invalid token ID or user ID'
            }
    
    def _validate_token_creation(self, amount: float, currency: str, user_id: str) -> Dict[str, Any]:
        """Validate token creation parameters"""
        
        # Check amount limits
        if amount < self.config['min_token_amount']:
            return {
                'valid': False,
                'reason': f"Amount below minimum: {self.config['min_token_amount']} {currency}"
            }
        
        if amount > self.config['max_token_amount']:
            return {
                'valid': False,
                'reason': f"Amount exceeds maximum: {self.config['max_token_amount']} {currency}"
            }
        
        # Check currency support
        if currency not in self.config['supported_currencies']:
            return {
                'valid': False,
                'reason': f"Currency {currency} not supported for tokens"
            }
        
        # Check user ID
        if not user_id:
            return {
                'valid': False,
                'reason': 'User ID required for token creation'
            }
        
        return {
            'valid': True,
            'reason': 'Token creation parameters valid'
        }
    
    def _generate_token_id(self) -> str:
        """Generate a secure token ID"""
        prefix = "TOKEN"
        timestamp = datetime.utcnow().strftime("%Y%m%d%H%M%S")
        random_part = secrets.token_hex(8).upper()
        return f"{prefix}_{timestamp}_{random_part}"
    
    def _generate_token_qr_code(self, token_id: str) -> str:
        """Generate QR code data for token"""
        # In production, this would generate actual QR code
        qr_data = {
            'type': 'offline_token',
            'token_id': token_id,
            'version': '1.0'
        }
        return f"pinkpay://token/{token_id}"
    
    def validate_input(self, data: Dict[str, Any]) -> bool:
        """Validate input data for token operations"""
        operation = data.get('token_operation', 'validate')
        
        if operation == 'create':
            required_fields = ['user_id', 'amount']
        elif operation in ['redeem', 'validate']:
            required_fields = ['token_id']
        elif operation == 'cancel':
            required_fields = ['token_id', 'user_id']
        else:
            return True  # No validation needed for other operations
        
        for field in required_fields:
            if field not in data:
                return False
        
        return True
    
    async def cleanup_expired_tokens(self) -> Dict[str, Any]:
        """Cleanup expired tokens (background task)"""
        if not self.config['auto_cleanup_expired']:
            return {'cleaned': 0, 'reason': 'Auto cleanup disabled'}
        
        # In production, this would query and update database
        # Mock cleanup operation
        await asyncio.sleep(0.2)
        
        return {
            'cleaned': 5,  # Mock: cleaned 5 expired tokens
            'cleanup_timestamp': datetime.utcnow().isoformat()
        } 
Token Handler Plugin for PinkPay Payment Switch
Manages offline payment tokens for situations with poor connectivity
"""

import asyncio
import secrets
from datetime import datetime, timedelta
from typing import Dict, Any, Optional
from .plugin_manager import BasePlugin

class TokenHandlerPlugin(BasePlugin):
    """Token Handler Plugin for offline payment tokens"""
    
    def __init__(self):
        super().__init__()
        self.name = "Token Handler"
        self.version = "1.0.0"
        self.description = "Offline payment token management and redemption"
        self.critical = False
        
        self.config = {
            'token_expiry_hours': 24,
            'max_token_amount': 1000.0,  # MYR
            'min_token_amount': 5.0,     # MYR
            'max_tokens_per_user': 5,
            'supported_currencies': ['MYR', 'USD', 'SGD'],
            'auto_cleanup_expired': True
        }
    
    async def execute(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Execute token handling logic"""
        try:
            operation = data.get('token_operation', 'validate')
            
            if operation == 'create':
                return await self._create_token(data)
            elif operation == 'redeem':
                return await self._redeem_token(data)
            elif operation == 'validate':
                return await self._validate_token(data)
            elif operation == 'cancel':
                return await self._cancel_token(data)
            else:
                return {
                    'success': True,
                    'data': {
                        'token_operation': 'none',
                        'token_required': False
                    }
                }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Token handling failed: {str(e)}",
                'critical': False
            }
    
    async def _create_token(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new offline token"""
        try:
            user_id = data.get('user_id')
            amount = float(data.get('amount', 0))
            currency = data.get('currency', 'MYR')
            
            # Validate token creation
            validation_result = self._validate_token_creation(amount, currency, user_id)
            if not validation_result['valid']:
                return {
                    'success': False,
                    'error': validation_result['reason'],
                    'critical': False
                }
            
            # Generate token
            token_data = await self._generate_offline_token(user_id, amount, currency)
            
            # In production, this would save to database
            # For demo, we'll simulate the database operation
            await asyncio.sleep(0.1)
            
            return {
                'success': True,
                'data': {
                    'token_created': True,
                    'token_data': token_data,
                    'token_operation': 'create'
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Token creation failed: {str(e)}",
                'critical': False
            }
    
    async def _redeem_token(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Redeem an offline token"""
        try:
            token_id = data.get('token_id')
            
            if not token_id:
                return {
                    'success': False,
                    'error': "Token ID required for redemption",
                    'critical': False
                }
            
            # Validate and redeem token
            redemption_result = await self._validate_and_redeem_token(token_id)
            
            return {
                'success': True,
                'data': {
                    'token_redeemed': redemption_result['valid'],
                    'redemption_data': redemption_result,
                    'token_operation': 'redeem'
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Token redemption failed: {str(e)}",
                'critical': False
            }
    
    async def _validate_token(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Validate an offline token without redeeming"""
        try:
            token_id = data.get('token_id')
            
            if not token_id:
                return {
                    'success': True,
                    'data': {
                        'token_valid': False,
                        'token_operation': 'validate',
                        'reason': 'No token provided'
                    }
                }
            
            # Validate token
            validation_result = await self._check_token_validity(token_id)
            
            return {
                'success': True,
                'data': {
                    'token_valid': validation_result['valid'],
                    'validation_data': validation_result,
                    'token_operation': 'validate'
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Token validation failed: {str(e)}",
                'critical': False
            }
    
    async def _cancel_token(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Cancel an offline token"""
        try:
            token_id = data.get('token_id')
            user_id = data.get('user_id')
            
            # Validate cancellation
            cancellation_result = await self._cancel_user_token(token_id, user_id)
            
            return {
                'success': True,
                'data': {
                    'token_cancelled': cancellation_result['cancelled'],
                    'cancellation_data': cancellation_result,
                    'token_operation': 'cancel'
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Token cancellation failed: {str(e)}",
                'critical': False
            }
    
    async def _generate_offline_token(self, user_id: str, amount: float, currency: str) -> Dict[str, Any]:
        """Generate a new offline token"""
        
        # Generate secure token ID
        token_id = self._generate_token_id()
        
        # Calculate expiry
        expires_at = datetime.utcnow() + timedelta(hours=self.config['token_expiry_hours'])
        
        # Create token data
        token_data = {
            'token_id': token_id,
            'user_id': user_id,
            'amount': amount,
            'currency': currency,
            'status': 'active',
            'created_at': datetime.utcnow().isoformat(),
            'expires_at': expires_at.isoformat(),
            'creation_location': 'unknown',  # Could be enhanced with geolocation
            'qr_code': self._generate_token_qr_code(token_id)
        }
        
        return token_data
    
    async def _validate_and_redeem_token(self, token_id: str) -> Dict[str, Any]:
        """Validate and redeem a token"""
        
        # Simulate database lookup
        await asyncio.sleep(0.1)
        
        # Mock token data (in production, fetch from database)
        mock_tokens = {
            'TOKEN_12345': {
                'token_id': 'TOKEN_12345',
                'user_id': 'user_001',
                'amount': 50.0,
                'currency': 'MYR',
                'status': 'active',
                'expires_at': (datetime.utcnow() + timedelta(hours=12)).isoformat()
            }
        }
        
        token_data = mock_tokens.get(token_id)
        
        if not token_data:
            return {
                'valid': False,
                'reason': 'Token not found',
                'redeemed': False
            }
        
        # Check if token is still valid
        if token_data['status'] != 'active':
            return {
                'valid': False,
                'reason': f"Token status is {token_data['status']}",
                'redeemed': False
            }
        
        # Check expiry
        expires_at = datetime.fromisoformat(token_data['expires_at'].replace('Z', '+00:00'))
        if datetime.utcnow() > expires_at.replace(tzinfo=None):
            return {
                'valid': False,
                'reason': 'Token has expired',
                'redeemed': False
            }
        
        # Token is valid, mark as redeemed
        token_data['status'] = 'redeemed'
        token_data['redeemed_at'] = datetime.utcnow().isoformat()
        
        return {
            'valid': True,
            'redeemed': True,
            'token_data': token_data,
            'redemption_timestamp': datetime.utcnow().isoformat()
        }
    
    async def _check_token_validity(self, token_id: str) -> Dict[str, Any]:
        """Check if a token is valid without redeeming it"""
        
        # Simulate database lookup
        await asyncio.sleep(0.05)
        
        # Mock validation (in production, query database)
        if token_id.startswith('TOKEN_'):
            return {
                'valid': True,
                'reason': 'Token is valid and active',
                'amount': 50.0,
                'currency': 'MYR',
                'expires_in_hours': 12
            }
        else:
            return {
                'valid': False,
                'reason': 'Invalid token format'
            }
    
    async def _cancel_user_token(self, token_id: str, user_id: str) -> Dict[str, Any]:
        """Cancel a user's token"""
        
        # Simulate database operation
        await asyncio.sleep(0.1)
        
        # Mock cancellation logic
        if token_id and user_id:
            return {
                'cancelled': True,
                'reason': 'Token cancelled successfully',
                'cancellation_timestamp': datetime.utcnow().isoformat()
            }
        else:
            return {
                'cancelled': False,
                'reason': 'Invalid token ID or user ID'
            }
    
    def _validate_token_creation(self, amount: float, currency: str, user_id: str) -> Dict[str, Any]:
        """Validate token creation parameters"""
        
        # Check amount limits
        if amount < self.config['min_token_amount']:
            return {
                'valid': False,
                'reason': f"Amount below minimum: {self.config['min_token_amount']} {currency}"
            }
        
        if amount > self.config['max_token_amount']:
            return {
                'valid': False,
                'reason': f"Amount exceeds maximum: {self.config['max_token_amount']} {currency}"
            }
        
        # Check currency support
        if currency not in self.config['supported_currencies']:
            return {
                'valid': False,
                'reason': f"Currency {currency} not supported for tokens"
            }
        
        # Check user ID
        if not user_id:
            return {
                'valid': False,
                'reason': 'User ID required for token creation'
            }
        
        return {
            'valid': True,
            'reason': 'Token creation parameters valid'
        }
    
    def _generate_token_id(self) -> str:
        """Generate a secure token ID"""
        prefix = "TOKEN"
        timestamp = datetime.utcnow().strftime("%Y%m%d%H%M%S")
        random_part = secrets.token_hex(8).upper()
        return f"{prefix}_{timestamp}_{random_part}"
    
    def _generate_token_qr_code(self, token_id: str) -> str:
        """Generate QR code data for token"""
        # In production, this would generate actual QR code
        qr_data = {
            'type': 'offline_token',
            'token_id': token_id,
            'version': '1.0'
        }
        return f"pinkpay://token/{token_id}"
    
    def validate_input(self, data: Dict[str, Any]) -> bool:
        """Validate input data for token operations"""
        operation = data.get('token_operation', 'validate')
        
        if operation == 'create':
            required_fields = ['user_id', 'amount']
        elif operation in ['redeem', 'validate']:
            required_fields = ['token_id']
        elif operation == 'cancel':
            required_fields = ['token_id', 'user_id']
        else:
            return True  # No validation needed for other operations
        
        for field in required_fields:
            if field not in data:
                return False
        
        return True
    
    async def cleanup_expired_tokens(self) -> Dict[str, Any]:
        """Cleanup expired tokens (background task)"""
        if not self.config['auto_cleanup_expired']:
            return {'cleaned': 0, 'reason': 'Auto cleanup disabled'}
        
        # In production, this would query and update database
        # Mock cleanup operation
        await asyncio.sleep(0.2)
        
        return {
            'cleaned': 5,  # Mock: cleaned 5 expired tokens
            'cleanup_timestamp': datetime.utcnow().isoformat()
        } 

File: backend/src/plugins/token-handler.js
const config = require('../utils/config');
const logger = require('../utils/logger');
const OfflineToken = require('../models/OfflineToken');
const { v4: uuidv4 } = require('uuid');

class TokenHandlerPlugin {
  constructor() {
    this.name = 'token-handler';
    this.version = '1.0.0';
    this.description = 'Offline payment token generation and management';

    // Token configuration
    this.tokenConfig = {
      defaultExpiryHours: 24, // 24 hours default
      maxExpiryHours: 168, // 1 week maximum
      minAmount: 1.0,
      maxAmount: 1000.0,
      allowedMerchantTypes: ['retail', 'restaurant', 'grocery', 'transport'],
      restrictedMerchants: ['gambling', 'adult', 'crypto']
    };
  }

  /**
   * Check if plugin should be enabled for this transaction
   */
  isEnabled(transaction, context) {
    // Enable for offline token generation/redemption requests
    return transaction.type === 'offline' ||
      context.operation === 'generateToken' ||
      context.operation === 'redeemToken';
  }

  /**
   * Process token operations
   */
  async process(transaction, context) {
    const startTime = Date.now();

    try {
      logger.info(`Token Handler processing transaction ${transaction.txnId}`, {
        operation: context.operation,
        type: transaction.type,
        amount: transaction.amount
      });

      let result;

      switch (context.operation) {
        case 'generateToken':
          result = await this.generateOfflineToken(transaction, context);
          break;
        case 'redeemToken':
          result = await this.redeemOfflineToken(transaction, context);
          break;
        case 'validateToken':
          result = await this.validateOfflineToken(transaction, context);
          break;
        default:
          throw new Error(`Unknown token operation: ${context.operation}`);
      }

      result.processingTime = Date.now() - startTime;

      logger.info(`Token operation completed for transaction ${transaction.txnId}`, {
        operation: context.operation,
        success: result.success,
        tokenId: result.token?.id || context.token
      });

      return result;

    } catch (error) {
      logger.logError(error, {
        action: 'tokenOperation',
        txnId: transaction.txnId,
        operation: context.operation
      });

      return {
        success: false,
        action: 'failed',
        operation: context.operation,
        error: error.message,
        processingTime: Date.now() - startTime
      };
    }
  }

  /**
   * Generate offline payment token
   */
  async generateOfflineToken(transaction, context) {
    try {
      // Validate token generation request
      const validation = this.validateTokenGeneration(transaction, context);
      if (!validation.valid) {
        throw new Error(validation.reason);
      }

      // Calculate expiry time
      const expiryHours = context.expiryHours || this.tokenConfig.defaultExpiryHours;
      const expiresAt = new Date(Date.now() + expiryHours * 60 * 60 * 1000);

      // Generate unique token
      const tokenString = this.generateUniqueToken();

      // Prepare merchant restrictions
      const merchantRestrictions = this.prepareMerchantRestrictions(context);

      // Create offline token
      const offlineToken = await OfflineToken.create({
        token: tokenString,
        userId: transaction.userId,
        amount: transaction.amount,
        currency: transaction.currency,
        expiresAt,
        merchantRestrictions
      });

      // Update transaction with token information
      await transaction.updateStatus('completed', {
        offlineToken: offlineToken.token,
        tokenId: offlineToken.id,
        expiresAt: offlineToken.expiresAt
      });

      return {
        success: true,
        action: 'generated',
        operation: 'generateToken',
        token: {
          id: offlineToken.id,
          token: offlineToken.token,
          amount: offlineToken.amount,
          currency: offlineToken.currency,
          expiresAt: offlineToken.expiresAt,
          merchantRestrictions: offlineToken.merchantRestrictions
        },
        transaction: {
          status: 'completed',
          metadata: {
            offlineToken: offlineToken.token,
            tokenId: offlineToken.id
          }
        }
      };

    } catch (error) {
      logger.logError(error, {
        action: 'generateOfflineToken',
        txnId: transaction.txnId
      });
      throw error;
    }
  }

  /**
   * Redeem offline payment token
   */
  async redeemOfflineToken(transaction, context) {
    try {
      if (!context.token) {
        throw new Error('Token is required for redemption');
      }

      // Find the offline token
      const offlineToken = await OfflineToken.findByToken(context.token);
      if (!offlineToken) {
        throw new Error('Invalid token');
      }

      // Validate token can be redeemed
      if (!offlineToken.canRedeem(context.merchantId)) {
        throw new Error('Token cannot be redeemed - expired, used, or merchant restrictions');
      }

      // Validate redemption amount
      if (transaction.amount && parseFloat(transaction.amount) !== parseFloat(offlineToken.amount)) {
        throw new Error(`Amount mismatch: expected ${offlineToken.amount}, got ${transaction.amount}`);
      }

      // Redeem the token
      await offlineToken.redeem(context.merchantId);

      // Update transaction with redemption details
      await transaction.updateStatus('completed', {
        redeemedToken: offlineToken.token,
        originalAmount: offlineToken.amount,
        redemptionMerchant: context.merchantId,
        redemptionTime: new Date().toISOString()
      });

      return {
        success: true,
        action: 'redeemed',
        operation: 'redeemToken',
        token: {
          id: offlineToken.id,
          token: offlineToken.token,
          amount: offlineToken.amount,
          currency: offlineToken.currency,
          status: 'used',
          redemptionTime: offlineToken.usedAt
        },
        redemption: {
          merchantId: context.merchantId,
          amount: offlineToken.amount,
          currency: offlineToken.currency
        },
        transaction: {
          status: 'completed',
          amount: offlineToken.amount,
          currency: offlineToken.currency
        }
      };

    } catch (error) {
      logger.logError(error, {
        action: 'redeemOfflineToken',
        txnId: transaction.txnId,
        token: context.token
      });
      throw error;
    }
  }

  /**
   * Validate offline payment token
   */
  async validateOfflineToken(transaction, context) {
    try {
      if (!context.token) {
        throw new Error('Token is required for validation');
      }

      // Find the offline token
      const offlineToken = await OfflineToken.findByToken(context.token);
      if (!offlineToken) {
        return {
          success: false,
          action: 'invalid',
          operation: 'validateToken',
          reason: 'Token not found'
        };
      }

      // Check token validity
      const isValid = offlineToken.canRedeem(context.merchantId);
      const validationDetails = {
        exists: true,
        expired: offlineToken.isExpired(),
        used: offlineToken.status === 'used',
        cancelled: offlineToken.status === 'cancelled',
        merchantAllowed: !context.merchantId || !offlineToken.merchantRestrictions.allowedMerchants ||
          offlineToken.merchantRestrictions.allowedMerchants.includes(context.merchantId),
        merchantBlocked: context.merchantId && offlineToken.merchantRestrictions.blockedMerchants &&
          offlineToken.merchantRestrictions.blockedMerchants.includes(context.merchantId)
      };

      return {
        success: true,
        action: isValid ? 'valid' : 'invalid',
        operation: 'validateToken',
        token: {
          id: offlineToken.id,
          amount: offlineToken.amount,
          currency: offlineToken.currency,
          status: offlineToken.status,
          expiresAt: offlineToken.expiresAt,
          isExpired: offlineToken.isExpired()
        },
        validation: {
          isValid,
          details: validationDetails
        }
      };

    } catch (error) {
      logger.logError(error, {
        action: 'validateOfflineToken',
        token: context.token
      });
      throw error;
    }
  }

  /**
   * Validate token generation request
   */
  validateTokenGeneration(transaction, context) {
    // Check amount limits
    if (transaction.amount < this.tokenConfig.minAmount) {
      return {
        valid: false,
        reason: `Amount too low: minimum ${this.tokenConfig.minAmount}`
      };
    }

    if (transaction.amount > this.tokenConfig.maxAmount) {
      return {
        valid: false,
        reason: `Amount too high: maximum ${this.tokenConfig.maxAmount}`
      };
    }

    // Check expiry limits
    const expiryHours = context.expiryHours || this.tokenConfig.defaultExpiryHours;
    if (expiryHours > this.tokenConfig.maxExpiryHours) {
      return {
        valid: false,
        reason: `Expiry too long: maximum ${this.tokenConfig.maxExpiryHours} hours`
      };
    }

    // Check merchant restrictions
    if (context.merchantType && this.tokenConfig.restrictedMerchants.includes(context.merchantType)) {
      return {
        valid: false,
        reason: `Restricted merchant type: ${context.merchantType}`
      };
    }

    return { valid: true };
  }

  /**
   * Generate unique token string
   */
  generateUniqueToken() {
    const timestamp = Date.now().toString(36);
    const random = Math.random().toString(36).substring(2, 8);
    const uuid = uuidv4().substring(0, 8);

    return `OT-${timestamp}-${random}-${uuid}`.toUpperCase();
  }

  /**
   * Prepare merchant restrictions from context
   */
  prepareMerchantRestrictions(context) {
    const restrictions = {};

    if (context.allowedMerchants && Array.isArray(context.allowedMerchants)) {
      restrictions.allowedMerchants = context.allowedMerchants;
    }

    if (context.blockedMerchants && Array.isArray(context.blockedMerchants)) {
      restrictions.blockedMerchants = context.blockedMerchants;
    }

    if (context.merchantType && this.tokenConfig.allowedMerchantTypes.includes(context.merchantType)) {
      restrictions.allowedTypes = [context.merchantType];
    }

    if (context.locationRestrictions) {
      restrictions.location = context.locationRestrictions;
    }

    return restrictions;
  }

  /**
   * Get token statistics
   */
  async getTokenStats(period = '24h') {
    try {
      return await OfflineToken.getStats(period);
    } catch (error) {
      logger.logError(error, { action: 'getTokenStats', period });
      throw error;
    }
  }

  /**
   * Cleanup expired tokens
   */
  async cleanupExpiredTokens() {
    try {
      const count = await OfflineToken.cleanupExpired();
      logger.info(`Cleaned up ${count} expired tokens`);

      return {
        success: true,
        action: 'cleanup',
        expiredTokensRemoved: count
      };
    } catch (error) {
      logger.logError(error, { action: 'cleanupExpiredTokens' });
      throw error;
    }
  }

  /**
   * Get token configuration
   */
  getTokenConfiguration() {
    return {
      config: this.tokenConfig,
      version: this.version,
      supportedOperations: ['generateToken', 'redeemToken', 'validateToken']
    };
  }

  /**
   * Update token configuration
   */
  updateTokenConfiguration(newConfig) {
    this.tokenConfig = { ...this.tokenConfig, ...newConfig };
    logger.info('Token configuration updated', newConfig);
  }

  /**
   * Check if this plugin is critical
   */
  isCritical(transaction, context) {
    // Token operations are critical for offline payments
    return transaction.type === 'offline' || context.operation?.includes('Token');
  }
}

module.exports = TokenHandlerPlugin; 